block/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -72,16 +83,67 @@ retrievalWorker keepAliveTimer = worker outs (retrievalWorkerImpl keepAliveTimer
-    handleAll mainLoopE $ do
-        logInfo "Starting retrievalWorker loop"
-        mainLoop
+    => Timer -> Maybe Text -> SendActions m -> m ()
+retrievalWorkerImpl keepAliveTimer blockStorageMirror SendActions {..} = do
+    handleAll tryDownloadE $ do
+        epoch <- view epochIndexL <$> DB.getTipHeader
+        whenJust blockStorageMirror $ \dumpUrl ->
+            tryDownload epoch (toString dumpUrl)
+    handleAll recoveryLoopE $ do
+        logInfo "Starting recovery loop"
+        recoveryLoop
   where
-    mainLoop = do
+    -- | Try to download and apply a blockchain dump, starting from the
+    -- given epoch.
+    tryDownload epoch dumpUrl = whenNothingM_ getCurrentSlot $ do
+        let epochUrl = dumpUrl <//> format "epoch{}.cbor.lzma"
+                                           (toInteger epoch)
+        logInfo ("Downloading blockchain dump for epoch "+|epoch|+
+                 " from "+|epochUrl|+"")
+        isSuccess <- modifyStateLock HighPriority "tryDownload" $ \tip -> do
+            slot <- getEpochOrSlot <$> DB.getTipHeader
+            -- When we're already in the epoch we're going to be downloading
+            -- blocks from, it's possible that we're in slot 10000 or
+            -- something and we'll have to skip 10000 blocks before we can
+            -- show any progress. In this case let's warn about it.
+            whenRight (unEpochOrSlot slot) $ \slotId ->
+                when (slotId ^. epochIndexL == epoch) $
+                    logDebug ("Going to download blocks and skip until slot "
+                              +|view siSlotL slotId|+", this may take " <>
+                              "some time")
+            let applyDump
+                    = decodeBlockDumpC
+                   .| (C.dropWhileC ((/= tip) . view prevBlockL)
+                   >> C.transPipe lift (verifyAndApplyBlocksC True))
+            -- We use Data.Conduit.Async to keep a buffer of downloaded
+            -- blocks and apply them in parallel with downloading them. The
+            -- buffer has size 'epochSlots' because that's the maximum
+            -- amount of blocks we can have in an epoch
+            request <- Http.parseRequest epochUrl
+            res <- C.runResourceT $
+                Conduit.Async.buffer (fromIntegral epochSlots)
+                    (Http.httpSource request Http.getResponseBody)
+                    applyDump
+            -- TODO: abort if downloading is too slow and we haven't gotten
+            -- any blocks in e.g. last minute
+            case res of
+                Left err -> do
+                    logError ("Failed to apply the dump: "+|err|+"")
+                    pure (tip, False)
+                Right newTip ->
+                    pure (newTip, True)
+        when isSuccess $ tryDownload (succ epoch) dumpUrl
+    tryDownloadE e =
+        reportOrLogE "retrievalWorker tryDownloadE: error caught " e
+
+    recoveryLoop = do
```

Comment:
It's not a recovery loop, it's a block retrieval worker loop, it doesn't only do recovery, also incremental "continues" synchronization.

Reply:
> also incremental "continues" synchronization

What does it mean? I can't parse it.

Reply:
See `CHContinues` and `brtContinues`. I mean that this loop is also responsible for "handle the newest header that has just been propagated".